ca/ca.go

```diff
@@ -659,8 +663,70 @@ func (ca *CertificateAuthorityImpl) generateOCSPAndStoreCertificate(
 		// changes here, you should make sure they are reflected in orphan-finder.
 		ca.log.AuditErrf("Failed RPC to store at SA, orphaning certificate: serial=[%s] cert=[%s] err=[%v], regID=[%d], orderID=[%d]",
 			core.SerialToString(serialBigInt), hex.EncodeToString(certDER), err, regID, orderID)
+		if ca.orphanQueue != nil {
+			ca.queueOrphan(&orphanedCert{
+				DER:      certDER,
+				OCSPResp: ocspResp,
+				RegID:    regID,
+			})
+		}
 		return core.Certificate{}, err
 	}
 
 	return core.Certificate{DER: certDER}, nil
 }
+
+type orphanedCert struct {
+	DER      []byte
+	OCSPResp []byte
+	RegID    int64
+}
+
+func (ca *CertificateAuthorityImpl) queueOrphan(o *orphanedCert) {
+	_, err := ca.orphanQueue.EnqueueObject(o)
+	if err != nil {
+		ca.log.Err(fmt.Sprintf("failed to queue orphan for integration: %s", err))
+	}
+}
+
+// OrphanIntegrationLoop runs a loop executing integrateOrphans and then waiting a minute.
+// It is split out into a separate function called directly by boulder-ca in order to make
+// testing the orphan queue functionality somewhat more simple.
+func (ca *CertificateAuthorityImpl) OrphanIntegrationLoop() {
+	for {
+		err := ca.integrateOrphans()
+		if err != nil && err != goque.ErrEmpty {
+			ca.log.Err(fmt.Sprintf("failed to integrate orphaned certs: %s", err))
+		}
+		time.Sleep(time.Minute)
+	}
+}
+
+func (ca *CertificateAuthorityImpl) integrateOrphans() error {
+	item, err := ca.orphanQueue.Peek()
+	if err != nil {
+		if err == goque.ErrEmpty {
+			return goque.ErrEmpty
+		}
+		return fmt.Errorf("failed to peek into orphan queue: %s", err)
+	}
+	var orphan orphanedCert
+	err = item.ToObject(&orphan)
+	if err != nil {
+		return fmt.Errorf("failed to marshal orphan: %s", err)
+	}
+	cert, err := x509.ParseCertificate(orphan.DER)
+	if err != nil {
+		return fmt.Errorf("failed to parse orphan: %s", err)
+	}
+	issued := cert.NotBefore.Add(-ca.backdate)
+	_, err = ca.sa.AddCertificate(context.Background(), orphan.DER, orphan.RegID, orphan.OCSPResp, &issued)
+	if err != nil && !strings.HasPrefix(err.Error(), "Error 1062: Duplicate entry") {
```

Comment:
Could we return a concrete error type for this case from `sa.AddCertificate` instead of having to check the `err.Error()`? That seems slightly less brittle.

Reply:
@rolandshoemaker Ping on this?

Reply:
Oops, missed this, will do.